packages/components/src/modal/test/index.tsx

```diff
@@ -236,4 +236,144 @@ describe( 'Modal', () => {
 			screen.getByText( 'A sweet button', { selector: 'button' } )
 		).toBeInTheDocument();
 	} );
+
+	describe( 'Focus handling', () => {
+		let originalGetClientRects: () => DOMRectList;
+
+		beforeEach( () => {
+			/**
+			 * The test environment does not have a layout engine, so we need to mock
+			 * the getClientRects method. This ensures that the focusable elements can be
+			 * found by the `focusOnMount` logic which depends on layout information
+			 * to determine if the element is visible or not.
+			 * See https://github.com/WordPress/gutenberg/blob/trunk/packages/dom/src/focusable.js#L55-L61.
+			 */
+			// @ts-expect-error We're not trying to comply to the DOM spec, only mocking
+			window.HTMLElement.prototype.getClientRects = function () {
+				return [ 'trick-jsdom-into-having-size-for-element-rect' ];
+			};
+		} );
+
+		afterEach( () => {
+			// Restore original HTMLElement prototype.
+			// See beforeEach for details.
+			window.HTMLElement.prototype.getClientRects =
+				originalGetClientRects;
+		} );
+
+		it( 'should focus the first focusable element in the contents (if found) when `firstContentElement` passed as value for `focusOnMount` prop', async () => {
+			const user = userEvent.setup();
+
+			const FocusMountDemo = () => {
+				const [ isShown, setIsShown ] = useState( false );
+				return (
+					<>
+						<button onClick={ () => setIsShown( true ) }>📣</button>
+						{ isShown && (
+							<Modal
+								focusOnMount="firstContentElement"
+								onRequestClose={ () => setIsShown( false ) }
+							>
+								<p>Modal content</p>
+								<a
+									href="https://wordpress.org"
+									data-testid="first-focusable-element"
+								>
+									First Focusable Element
+								</a>
+
+								<a href="https://wordpress.org">
+									Another Focusable Element
+								</a>
+							</Modal>
+						) }
+					</>
+				);
+			};
```

Comment:
We can dry up the code a little bit by refactoring the React test component outside of the unit tests

<details>

<summary>Something like this</summary>

```diff
diff --git a/packages/components/src/modal/test/index.tsx b/packages/components/src/modal/test/index.tsx
index 0a3111a6df..9baf611034 100644
--- a/packages/components/src/modal/test/index.tsx
+++ b/packages/components/src/modal/test/index.tsx
@@ -13,6 +13,7 @@ import { useState } from '@wordpress/element';
  * Internal dependencies
  */
 import Modal from '../';
+import type { ModalProps } from '../types';
 
 const noop = () => {};
 
@@ -238,6 +239,32 @@ describe( 'Modal', () => {
 	} );
 
 	describe( 'Focus handling', () => {
+		const FocusDemo = ( {
+			focusOnMount,
+		}: Pick< ModalProps, 'focusOnMount' > ) => {
+			const [ isShown, setIsShown ] = useState( false );
+			return (
+				<>
+					<button onClick={ () => setIsShown( true ) }>📣</button>
+					{ isShown && (
+						<Modal
+							focusOnMount={ focusOnMount }
+							onRequestClose={ () => setIsShown( false ) }
+						>
+							<p>Modal content</p>
+							<a href="https://wordpress.org">
+								First Focusable Element
+							</a>
+
+							<a href="https://wordpress.org">
+								Another Focusable Element
+							</a>
+						</Modal>
+					) }
+				</>
+			);
+		};
+
 		let originalGetClientRects: () => DOMRectList;
 
 		beforeEach( () => {
@@ -264,34 +291,7 @@ describe( 'Modal', () => {
 		it( 'should focus the first focusable element in the contents (if found) when `firstContentElement` passed as value for `focusOnMount` prop', async () => {
 			const user = userEvent.setup();
 
-			const FocusMountDemo = () => {
-				const [ isShown, setIsShown ] = useState( false );
-				return (
-					<>
-						<button onClick={ () => setIsShown( true ) }>📣</button>
-						{ isShown && (
-							<Modal
-								focusOnMount="firstContentElement"
-								onRequestClose={ () => setIsShown( false ) }
-							>
-								<p>Modal content</p>
-								<a
-									href="https://wordpress.org"
-									data-testid="first-focusable-element"
-								>
-									First Focusable Element
-								</a>
-
-								<a href="https://wordpress.org">
-									Another Focusable Element
-								</a>
-							</Modal>
-						) }
-					</>
-				);
-			};
-
-			render( <FocusMountDemo /> );
+			render( <FocusDemo focusOnMount="firstContentElement" /> );
 
 			const opener = screen.getByRole( 'button' );
 
@@ -306,33 +306,8 @@ describe( 'Modal', () => {
 
 		it( 'should focus the Modal dialog when `true` passed as value for `focusOnMount` prop', async () => {
 			const user = userEvent.setup();
-			const FocusMountDemo = () => {
-				const [ isShown, setIsShown ] = useState( false );
-				return (
-					<>
-						<button onClick={ () => setIsShown( true ) }>📣</button>
-						{ isShown && (
-							<Modal
-								focusOnMount
-								onRequestClose={ () => setIsShown( false ) }
-							>
-								<p>Modal content</p>
-								<a
-									href="https://wordpress.org"
-									data-testid="first-focusable-element"
-								>
-									First Focusable Element
-								</a>
 
-								<a href="https://wordpress.org">
-									Another Focusable Element
-								</a>
-							</Modal>
-						) }
-					</>
-				);
-			};
-			render( <FocusMountDemo /> );
+			render( <FocusDemo focusOnMount /> );
 
 			const opener = screen.getByRole( 'button' );
 
@@ -343,33 +318,8 @@ describe( 'Modal', () => {
 
 		it( 'should not move focus when `false` passed as value for `focusOnMount` prop', async () => {
 			const user = userEvent.setup();
-			const FocusMountDemo = () => {
-				const [ isShown, setIsShown ] = useState( false );
-				return (
-					<>
-						<button onClick={ () => setIsShown( true ) }>📣</button>
-						{ isShown && (
-							<Modal
-								focusOnMount={ false }
-								onRequestClose={ () => setIsShown( false ) }
-							>
-								<p>Modal content</p>
-								<a
-									href="https://wordpress.org"
-									data-testid="first-focusable-element"
-								>
-									First Focusable Element
-								</a>
 
-								<a href="https://wordpress.org">
-									Another Focusable Element
-								</a>
-							</Modal>
-						) }
-					</>
-				);
-			};
-			render( <FocusMountDemo /> );
+			render( <FocusDemo focusOnMount={ false } /> );
 
 			const opener = screen.getByRole( 'button' );
 
```
</details>

Reply:
✅ Refactor complete.